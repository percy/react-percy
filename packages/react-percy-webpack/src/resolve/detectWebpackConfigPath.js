import chalk from 'chalk';
import fs from 'fs';
import path from 'path';

export default function detectWebpackConfigPath(packageRoot, resolve = require.resolve) {
  const defaultWebpackPath = path.join(packageRoot, 'webpack.config.js');
  if (fs.existsSync(defaultWebpackPath)) {
    return defaultWebpackPath;
  }

  const packageJson = require(path.join(packageRoot, 'package.json'));
  if (
    (packageJson.dependencies && packageJson.dependencies['react-scripts']) ||
    (packageJson.devDependencies && packageJson.devDependencies['react-scripts']) ||
    (packageJson.optionalDependencies && packageJson.optionalDependencies['react-scripts'])
  ) {
    try {
      return resolve('react-scripts/config/webpack.config.dev');
    } catch (e) {
      // Webpack config is not in the expected location
    }
  }

  throw new Error(
    chalk.red('No webpack config file found. Use ') +
      chalk.red.bold('`react-percy --config <path>`') +
      chalk.red(' to specify the path of your webpack config.'),
  );
}
