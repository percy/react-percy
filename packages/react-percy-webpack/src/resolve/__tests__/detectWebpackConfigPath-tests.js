import detectWebpackConfigPath from '../detectWebpackConfigPath';
import fs from 'fs';

jest.mock('fs', () => ({
  existsSync: jest.fn(),
}));

const packageRoot = '/some/package/root';

let mockPackage;
jest.mock('/some/package/root/package.json', () => mockPackage, { virtual: true });

let exists;
let resolve;
let resolvePaths;

beforeEach(() => {
  jest.resetModules();

  exists = new Set();
  fs.existsSync.mockReset();
  fs.existsSync.mockImplementation(file => exists.has(file));

  resolvePaths = {};
  resolve = jest.fn(name => resolvePaths[name]);

  mockPackage = {};
});

it('returns path of webpack.config.js in package root when found', () => {
  exists.add('/some/package/root/webpack.config.js');

  const configPath = detectWebpackConfigPath(packageRoot, resolve);

  expect(configPath).toBe('/some/package/root/webpack.config.js');
});

it('returns path of react-scripts dev webpack.config.js when react-scripts is installed as a dependency', () => {
  mockPackage.dependencies = {
    'react-scripts': '^1.1.0',
  };
  resolvePaths['react-scripts/config/webpack.config.dev'] =
    '/some/package/root/node_modules/react-scripts/config/webpack.config.dev.js';

  const configPath = detectWebpackConfigPath(packageRoot, resolve);

  expect(configPath).toBe(
    '/some/package/root/node_modules/react-scripts/config/webpack.config.dev.js',
  );
});

it('returns path of react-scripts dev webpack.config.js when react-scripts is installed as a dev dependency', () => {
  mockPackage.devDependencies = {
    'react-scripts': '^1.1.0',
  };
  resolvePaths['react-scripts/config/webpack.config.dev'] =
    '/some/package/root/node_modules/react-scripts/config/webpack.config.dev.js';

  const configPath = detectWebpackConfigPath(packageRoot, resolve);

  expect(configPath).toBe(
    '/some/package/root/node_modules/react-scripts/config/webpack.config.dev.js',
  );
});

it('throws when there is no webpack config in the package root and it is not a create-react-app project', () => {
  expect(() => detectWebpackConfigPath(packageRoot, resolve)).toThrowErrorMatchingSnapshot();
});
