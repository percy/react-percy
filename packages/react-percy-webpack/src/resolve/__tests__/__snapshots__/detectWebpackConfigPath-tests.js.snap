// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`throws when there is no webpack config in the package root and it is not a create-react-app project 1`] = `"[31mNo webpack config file found. Use [39m[31m[1m\`react-percy --config <path>\`[22m[39m[31m to specify the path of your webpack config.[39m"`;
